### PR TITLE
feat(memoization): knapsack implementation

### PR DESCRIPTION
**Pull request by Naramsim**

### CODE REVIEW

File: lib/algo/memoization/0-1 Knapsack/index.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  if (currentItem !== -1) {
+    return currentItem;
+  }
+  const taking = (knapSack(items, currentIndex - 1, residualCapacity - items[currentIndex].volume, valuesMatrix) + items[currentIndex].profit) || 0;
+  const notTaking = knapSack(items, currentIndex - 1, residualCapacity, valuesMatrix) || 0;
+  currentItem = Math.max(taking, notTaking);
+  return currentItem;
+}
+
+// console.log(knapSack(items, items.length -1 , capacity)) //starting from last item
```
Comment (sarbbottam) [inline lib/algo/memoization/0-1 Knapsack/index.js]: please remove `console.log`

File: lib/algo/memoization/0-1 Knapsack/index.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    return Number.MIN_SAFE_INTEGER;
+  }
+  if (currentIndex < 0) {
+    return 0;
+  }
+  let currentItem = valuesMatrix[currentIndex][residualCapacity] || 0;
+  if (currentItem !== -1) {
+    return currentItem;
+  }
+  const taking = (knapSack(items, currentIndex - 1, residualCapacity - items[currentIndex].volume, valuesMatrix) + items[currentIndex].profit) || 0;
```
Comment (sarbbottam) [inline lib/algo/memoization/0-1 Knapsack/index.js]: please refactor this line.

const taking = (knapSack(items, currentIndex - 1, residualCapacity - items[currentIndex].volume, valuesMatrix) + items[currentIndex].profit) || 0;

You might not want to use `||` for default value, you can read [more at ](https://sarbbottam.github.io/javascript/2016/05/27/never-use-or-for-default-values/)

File: lib/algo/memoization/0-1 Knapsack/index.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  const notTaking = knapSack(items, currentIndex - 1, residualCapacity, valuesMatrix) || 0;
```
Comment (sarbbottam) [inline lib/algo/memoization/0-1 Knapsack/index.js]: You might not want to use `||` for default value, you can read [more at ](https://sarbbottam.github.io/javascript/2016/05/27/never-use-or-for-default-values/)

File: lib/algo/memoization/0-1 Knapsack/index.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
+function createMatrix(x, y) {
```
Comment (sarbbottam) [inline lib/algo/memoization/0-1 Knapsack/index.js]: Please rename the file to `knapsack/index.js`, from `0-1 Knapsack/index.js`.

File: lib/algo/memoization/test.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
+const chai = require('chai');
+const knapsack = require('./0-1 Knapsack');
+
+class Item {
+  constructor(profit, volume) {
+    this.profit = profit;
+    this.volume = volume;
+  }
+}
+
```
Comment (sarbbottam) [inline lib/algo/memoization/test.js:15]: I would not use a `class` for returning a plain object w/o any custom functionalities, You can read more about [`constructor vs object literal` on why](https://www.google.com/search?q=constructor%20vs%20object%20literal).

You can can use `function`

``` js
function getItem (profit, volume)  {
  return {
    profit,
    volume
  }
}
```

File: lib/algo/memoization/test.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    new Item(5, 4),
+    new Item(1, 5),
+    new Item(56, 23),
+    new Item(2, 6),
+    new Item(1, 5),
+    new Item(12, 8),
+    new Item(2, 3)
+  ];
+  const capacity = 7;
+  it('should solve knapsack', done => {
```
Comment (sarbbottam) [inline lib/algo/memoization/test.js]: Can there be a better title, for persons who don't know `knapsack`, it would be difficult to follow. Something like
`should return maximum total value for capacity 7`

File: lib/algo/memoization/test.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    new Item(12, 2),
+    new Item(5, 4),
+    new Item(1, 5),
+    new Item(56, 23),
+    new Item(2, 6),
+    new Item(1, 5),
+    new Item(12, 8),
+    new Item(2, 3)
+  ];
+  const capacity = 7;
```
Comment (sarbbottam) [inline lib/algo/memoization/test.js]: Please move `const items` and `const capacity` to `it` block, as they should vary for each test case.

If you consider item to be constant for all the test cases, please take it out of `describe`.

File: lib/algo/memoization/test.js (hunk 6 above)
Comment (sarbbottam) [inline lib/algo/memoization/test.js]: Why `done`? Isn't `knapsack` a `synchronous` execution?

File: lib/algo/memoization/0-1 Knapsack/index.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+  const matrix = [];
+  for (let i = 0; i < x; i++) {
+    matrix[i] = [...new Array(y)].map(() => -1);
+  }
+  return matrix;
+}
+
+function knapSack(items, currentIndex, residualCapacity, valuesMatrix = createMatrix(currentIndex + 1, residualCapacity + 1)) {
```
Comment (sarbbottam) [inline lib/algo/memoization/0-1 Knapsack/index.js]: initial `currentIndex` is `items.length - 1`. How about defaulting it?

``` js
function knapSack(items, residualCapacity, currentIndex = items.length - 1, valuesMatrix = createMatrix(currentIndex + 1, residualCapacity + 1)) {
```

**Comment by sarbbottam** [comment]
Please add a comment to the source code pointing to `http://www.geeksforgeeks.org/dynamic-programming-set-10-0-1-knapsack-problem/`

**Comment by sarbbottam** [comment]
Right now there are `6 commits`, [please quash them into one](https://github.com/ginatrapani/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit).

**Comment by Naramsim** [comment]
@sarbbottam, should I test the `createMatrix()` function?

**Comment by sarbbottam** [comment]
> should I test the createMatrix() function?

Not required, right now it is private not exposed, so it is by default getting covered via public interface.
Should you want to maintain it as a shared utility then you must.

File: lib/algo/memoization/knapsack/index.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    return 0;
+  }
+  let currentItem = valuesMatrix[currentIndex][residualCapacity];
+  if (currentItem !== -1) {
+    return currentItem;
+  }
+  let taking = (knapSack(items, residualCapacity - items[currentIndex].getVolume(), currentIndex - 1, valuesMatrix) + items[currentIndex].getProfit());
+  if (!taking) {
+    taking = 0;
+  }
```
Comment (sarbbottam) [inline lib/algo/memoization/knapsack/index.js:29]: Why is this required? `knapSack` will never return any `falsy` values other than `0`, which is by principle. 

Am I misinterpreting?

File: lib/algo/memoization/knapsack/index.js [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+  let notTaking = knapSack(items, residualCapacity, currentIndex - 1, valuesMatrix);
+  if (!notTaking) {
+    notTaking = 0;
+  }
```
Comment (sarbbottam) [inline lib/algo/memoization/knapsack/index.js:33]: Why is this required? `knapSack` will never return any `falsy` values other than `0`, which is by principle. 

Am I misinterpreting?

File: lib/algo/memoization/test.js [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+function Item(profit, volume) {
+  this.profit = profit;
+  this.volume = volume;
+
+  this.getProfit = function () {
+    return this.profit;
+  };
+  this.getVolume = function () {
+    return this.volume;
+  };
```
Comment (sarbbottam) [inline lib/algo/memoization/test.js:13]: `getProfit` & `getVolume` is not making any improvement as `profit` & `volume` is publicly available.

BTW, is `profit` & `volume` a standard construct? Or they can be generalized. Should they be generalized as `weights` and `values`

File: lib/algo/memoization/test.js [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+const expect = chai.expect;
+const items = [
+  new Item(12, 2),
+  new Item(5, 4),
+  new Item(1, 5),
+  new Item(56, 23),
+  new Item(2, 6),
+  new Item(1, 5),
+  new Item(12, 8),
+  new Item(2, 3)
```
Comment (sarbbottam) [inline lib/algo/memoization/test.js:25]: Don't use a constructor if you don't need it.

File: lib/algo/sort/test.js [hunk 13]
```diff
@@ -1,4 +1,6 @@
-const assert = require('chai').assert;
```
Comment (sarbbottam) [inline lib/algo/sort/test.js:1]: Please take this out of this PR and raise a separate one. [Address only one use case in one PR.](https://github.com/sarbbottam/js-ds/blob/master/contributing.md#pull-request)

File: package.json [hunk 14]
```diff
@@ -24,6 +24,7 @@
   "homepage": "https://github.com/sarbbottam/js-ds",
   "devDependencies": {
     "chai": "^3.5.0",
+    "chai-sorted": "^0.1.0",
```
Comment (sarbbottam) [inline package.json:27]: I guess this change is related to `sort` refactor.

Please take this out of this PR and raise a separate one. [Address only one use case in one PR.](https://github.com/sarbbottam/js-ds/blob/master/contributing.md#pull-request)

File: package.json [hunk 15]
```diff
@@ -34,6 +35,9 @@
       "mocha"
     ],
     "esnext": true,
-    "space": true
+    "space": true,
+    "rules": {
+      "linebreak-style": 0
```
Comment (sarbbottam) [inline package.json:40]: Are you on `windows`? What error did it cause w/o this? Probably I should add a `.gitattribute` to the repo.

**Comment by Naramsim** [comment]
Yes, I am on win, it says me that lines shuold be CL but it are CLRF.
